rust/lance-index/src/scalar/flat.rs

```diff
@@ -0,0 +1,240 @@
+            .clone())
+    }
+
+    // Note that there is no write/train method for flat index at the moment and so it isn't
+    // really possible for this method to be called.  If there was we assume it will write all
+    // data as a single batch named data.lance
+    async fn load(store: Arc<dyn IndexStore>) -> Result<Arc<Self>> {
+        let batches = store.open_index_file("data.lance").await?;
+        let batch = batches.read_record_batch(0).await?;
+        Ok(Arc::new(Self {
+            data: Arc::new(batch),
+        }))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use arrow_array::types::Int32Type;
+    use arrow_array::types::UInt64Type;
+    use datafusion_common::ScalarValue;
+    use lance_datagen::{array, gen, RowCount};
+
+    fn example_index() -> FlatIndex {
+        let batch = gen()
+            .col(
+                Some("values".to_string()),
+                array::cycle::<Int32Type>(vec![10, 100, 1000, 1234]),
+            )
+            .col(
+                Some("ids".to_string()),
+                array::cycle::<UInt64Type>(vec![5, 0, 3, 100]),
+            )
+            .into_batch_rows(RowCount::from(4))
+            .unwrap();
+
+        FlatIndex {
+            data: Arc::new(batch),
+        }
+    }
+
+    async fn check_index(query: &ScalarQuery, expected: &[u64]) {
+        let index = example_index();
+        let actual = index.search(query).await.unwrap();
+        let expected = UInt64Array::from_iter_values(expected.iter().copied());
+        assert_eq!(actual, expected);
+    }
+
+    #[tokio::test]
+    async fn test_equality() {
+        check_index(&ScalarQuery::Equals(ScalarValue::from(100)), &[0]).await;
+        check_index(&ScalarQuery::Equals(ScalarValue::from(10)), &[5]).await;
+        check_index(&ScalarQuery::Equals(ScalarValue::from(5)), &[]).await;
+    }
+
+    #[tokio::test]
+    async fn test_range() {
+        check_index(
+            &ScalarQuery::Range(
+                Bound::Included(ScalarValue::from(100)),
+                Bound::Excluded(ScalarValue::from(1234)),
+            ),
+            &[0, 3],
+        )
+        .await;
+        check_index(
+            &ScalarQuery::Range(Bound::Unbounded, Bound::Excluded(ScalarValue::from(1000))),
+            &[5, 0],
+        )
+        .await;
+        check_index(
+            &ScalarQuery::Range(Bound::Included(ScalarValue::from(0)), Bound::Unbounded),
+            &[5, 0, 3, 100],
+        )
+        .await;
+        check_index(
+            &ScalarQuery::Range(Bound::Included(ScalarValue::from(100000)), Bound::Unbounded),
+            &[],
+        )
+        .await;
+    }
+
+    #[tokio::test]
+    async fn test_is_in() {
+        check_index(
+            &ScalarQuery::IsIn(vec![
+                ScalarValue::from(100),
+                ScalarValue::from(1234),
+                ScalarValue::from(3000),
+            ]),
+            &[0, 100],
+        )
+        .await;
+    }
+}
```

Comment:
Perhaps it's also worth testing with floats and strings?

Reply:
Done.  There is a test for the different data types that tests a btree and, indirectly, the flat index.  It doesn't test the matching in great detail but mostly that should be handled by arrow-rs/datafusion